cli/internal/lockfile/pnpm_lockfile.go

```diff
@@ -477,17 +470,122 @@ func formatPnpmKeyV6(name string, version string) string {
-	if atIndex == -1 || len(key) == atIndex+1 {
-		return ""
+	dp := parseDepPath(key)
+	if dp.peerSuffix != "" {
+		return fmt.Sprintf("%s_%s", dp.version, dp.peerSuffix)
 	}
-	return key[atIndex+1:]
+	return dp.version
 }
 
 func getVersionFromKeyV6(key string) string {
-	atIndex := strings.LastIndex(key, "@")
-	if atIndex == -1 || len(key) == atIndex+1 {
+	oldKey := convertNewToOldDepPath(key)
+	dp := parseDepPath(oldKey)
+	if dp.peerSuffix != "" {
+		return dp.version + dp.peerSuffix
+	}
+	return dp.version
+}
+
+// Parsed representation of a pnpm lockfile key
+type depPath struct {
+	host       string
+	name       string
+	version    string
+	peerSuffix string
+}
+
+func parseDepPath(dependency string) depPath {
+	// See https://github.com/pnpm/pnpm/blob/185ab01adfc927ea23d2db08a14723bf51d0025f/packages/dependency-path/src/index.ts#L96
+	var dp depPath
+	parts := strings.Split(dependency, "/")
+	shift := func() string {
+		if len(parts) == 0 {
+			return ""
+		}
+		val := parts[0]
+		parts = parts[1:]
+		return val
+	}
+
+	isAbsolute := dependency[0] != '/'
+	// Skip leading '/'
+	if !isAbsolute {
+		shift()
+	}
+
+	if isAbsolute {
+		dp.host = shift()
+	}
+
+	if len(parts) == 0 {
+		return dp
+	}
+
+	if strings.HasPrefix(parts[0], "@") {
+		dp.name = fmt.Sprintf("%s/%s", shift(), shift())
+	} else {
+		dp.name = shift()
+	}
+
+	version := strings.Join(parts, "/")
+	if len(version) > 0 {
+		var peerSuffixIndex int
+		if strings.Contains(version, "(") && strings.HasSuffix(version, ")") {
+			// v6 encodes peers deps using (peer=version)
+			// also used to encode patches using (path_hash=hash)
+			peerSuffixIndex = strings.Index(version, "(")
+			dp.peerSuffix = version[peerSuffixIndex:]
+			dp.version = version[0:peerSuffixIndex]
+		} else {
+			// pre v6 uses _ to separate version from peer dependencies
+			// if a dependency is patched and has peer dependencies its version will
+			// be encoded as version_patchHash_peerDepsHash
+			peerSuffixIndex = strings.Index(version, "_")
+			if peerSuffixIndex != -1 {
+				dp.peerSuffix = version[peerSuffixIndex+1:]
+				dp.version = version[0:peerSuffixIndex]
+			}
+		}
+		if peerSuffixIndex == -1 {
+			dp.version = version
+		}
+	}
+
+	return dp
+}
+
+var _patchHashKey = "patch_hash="
+
+func (d depPath) patchHash() string {
+	if strings.HasPrefix(d.peerSuffix, "(") && strings.HasSuffix(d.peerSuffix, ")") {
+		for _, part := range strings.Split(d.peerSuffix, "(") {
+			if strings.HasPrefix(part, _patchHashKey) {
+				// drop the enclosing ')'
+				return part[len(_patchHashKey) : len(part)-1]
+			}
+		}
+		// no patch entry found
 		return ""
 	}
-	return key[atIndex+1:]
+
+	sepIndex := strings.Index(d.peerSuffix, "_")
+	if sepIndex != -1 {
+		return d.peerSuffix[:sepIndex]
+	}
+	// if a dependency just has a single suffix we can't tell if it's a patch or peer hash
+	// return it in case it's a patch hash
+	return d.peerSuffix
+}
+
+func convertNewToOldDepPath(newPath string) string {
```

Comment:
Is this logic from the same SHA as above? Can you add a comment to make that super-clear?